indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorSpec.java

```diff
@@ -202,6 +205,55 @@ public boolean isSuspended()
     return suspended;
   }
 
+  /**
+   * Default implementation that prevents unsupported evolution of the supervisor spec
+   * <ul>
+   * <li>You cannot migrate between types of supervisors.</li>
+   * <li>You cannot change the input source stream of a running supervisor.</li>
+   * </ul>
+   * @param proposedSpec the proposed supervisor spec
+   * @throws DruidException if the proposed spec update is not allowed
+   */
+  @Override
+  public void validateSpecUpdateTo(SupervisorSpec proposedSpec) throws DruidException
+  {
+    if (!(proposedSpec instanceof SeekableStreamSupervisorSpec)) {
+      throw InvalidInput.exception(
+          StringUtils.format("Cannot evolve to [%s] from [%s]", getClass().getName(), proposedSpec.getClass().getName())
+      );
+    }
+    SeekableStreamSupervisorSpec other = (SeekableStreamSupervisorSpec) proposedSpec;
+    if (this.getSource() == null || other.getSource() == null) {
+      // Not likely to happen, but covering just in case.
+      throw InvalidInput.exception("Cannot consider SeekableStreamSupervisorSpec evolution when one or both of "
+                                   + "the specs have not provided an input source stream in the IOConfig.");
+    }
+
+    if (!this.getSource().equals(other.getSource())) {
+      throw InvalidInput.exception(getIllegalInputSourceUpdateErrorMessage(this.getSource(), other.getSource()));
+    }
+  }
+
   protected abstract SeekableStreamSupervisorSpec toggleSuspend(boolean suspend);
 
+  /**
+   * Returns an error message for illegal update of the input source stream.
+   * <p>
+   * This is a reasonable default message, but subclasses may override it to provide more domain specific terminology.
+   * </p>
+   * @param existingSource The existing input source stream
+   * @param proposedSource The proposed input source stream
+   * @return A formatted error message
+   */
+  protected String getIllegalInputSourceUpdateErrorMessage(String existingSource, String proposedSource)
```

Comment:
This method shouldn't be a part of the contract of this interface.
I would suggest just making the string a constant and formatting it when needed.